### ISSUE TITLE
Sudden performance drop when value_grad_norm returns Inf

### ISSUE THREAD

**Issue by tungnd1705** [danijar/dreamer#32]
Hi, when I run Dreamer for the finger spin task, I encountered a pretty strange problem. The algorithm was working well until the value_grad_norm returns Inf and the performance dropped suddenly. I know that you are using LossScaleOptimizer in the code base, so it should not apply the gradient if the grad norm is infinite.
Do you know why this happened? And do you have any suggestions for fixing this problem? Thank you!
![image](https://user-images.githubusercontent.com/36795321/88481283-a17a4900-cf84-11ea-9380-2d3ebe1ddc9e.png)

**Comment by danijar** [comment]
Are you training in float16 or float32?

**Comment by tungnd1705** [comment]
I'm training in float16

**Comment by danijar** [comment]
Could you give it a try in float32 to see if you still run into instabilities? If so, it's probably not something the LossScaleOptimizer can protect against.

**Comment by tungnd1705** [comment]
Thanks, I'll give it a try. So these are only numerical issues? Do you have any suggestions to avoid this problem while using float16?

**Comment by danijar** [comment]
I don't know what the problem is yet, so we'll have to see based on your float32 run.

**Comment by tungnd1705** [comment]
I've tried and running in float32 did not have that problem. What do you think the source of problem is?

**Comment by danijar** [comment]
Interesting. It seems to be related to training in lower precision then. You're right that the LossScaleOptimizer ignores gradients that are not finite to adjust the loss scale and avoid underflow/overflow over time.

One idea I have is that if too many gradients are ignored in a row, the ratio of actor and value updates changes so one could overfit to the other. This is just a vague guess though.

If you can afford training in float32 I would just recommend that. Otherwise it would be good to run a second experiment with float16 to see how often this happens.

**Comment by danijar** [comment]
Closing due to inactivity but feel free to comment and I'll reopen.
